<a href="https://colab.research.google.com/github/adalves-ufabc/2022.Q2-PLN/blob/main/2022_Q2_PLN_Notebook_13.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Processamento de Linguagem Natural [2022.Q2]**
Prof. Alexandre Donizeti Alves


# **Classificação de Textos**

In [ ]:
# importando as bibliotecas
import pandas as pd
import numpy as np

### **Conjunto de dados de coleta de *spam* por SMS**

O ***SMS Spam Collection*** é um conjunto público de mensagens com SMS que foram coletadas para pesquisas de *spam* em telefones celulares.

O dataset pode ser acessado no seguinte [link](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection).



In [ ]:
df = pd.read_csv('/content/spam.csv',encoding='latin-1')

In [ ]:
# checando o shape do dataset
df.shape

(5572, 5)

In [ ]:
# checando o head do dataset
# ham eh o rotulo para mensagens que nao sao spam
df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [ ]:
df.tail()

v1                                                 v2 Unnamed: 2  \
5567  spam  This is the 2nd time we have tried 2 contact u...        NaN   
5568   ham              Will Ì_ b going to esplanade fr home?        NaN   
5569   ham  Pity, * was in mood for that. So...any other s...        NaN   
5570   ham  The guy did some bitching but I acted like i'd...        NaN   
5571   ham                         Rofl. Its true to its name        NaN   

     Unnamed: 3 Unnamed: 4  
5567        NaN        NaN  
5568        NaN        NaN  
5569        NaN        NaN  
5570        NaN        NaN  
5571        NaN        NaN

In [ ]:
# renomeando as colunas
df = df.rename(columns={'v1': 'label', 'v2': 'message'})
df.head()

label                                            message Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [ ]:
# examinando a distribuicao
df.label.value_counts()

ham     4825
spam     747
Name: label, dtype: int64

In [ ]:
# hot-encoding a coluna de label
df['label_num']=df.label.map({"ham":0,"spam":1})

# checando a conversao
df.head()

label                                            message Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  label_num  
0        NaN        NaN          0  
1        NaN        NaN          0  
2        NaN        NaN          1  
3        NaN        NaN          0  
4        NaN        NaN          0

In [ ]:
# estabelecendo x e y
x = df.message
y = df.label_num
print(x.shape)
print(y.shape)

(5572,)
(5572,)


In [ ]:
# separar x e y em treino e teste
from sklearn.model_selection import train_test_split

# test_size [default]: 0.25
x_train,x_test,y_train,y_test = train_test_split(x,y,random_state=1)

print(x_train.shape)
print(x_test.shape)

(4179,)
(1393,)


### **Vetorizando os dados**

In [ ]:
# import o CountVectorizer
from sklearn.feature_extraction.text import CountVectorizer

vect = CountVectorizer()

# criando a matriz de termos
x_train_dtm = vect.fit_transform(x_train)
x_train_dtm

<4179x7496 sparse matrix of type '<class 'numpy.int64'>'
	with 55614 stored elements in Compressed Sparse Row format>

In [ ]:
# transformar o conjunto de teste na matriz
x_test_dtm = vect.transform(x_test)
x_test_dtm

<1393x7496 sparse matrix of type '<class 'numpy.int64'>'
	with 17010 stored elements in Compressed Sparse Row format>

### **Classificador *Naive Bayes*** 

In [ ]:
# importando bayes
from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB()

In [ ]:
# treiando o modelo usando o x _train_dtm
%time nb.fit(x_train_dtm, y_train)

CPU times: user 6.07 ms, sys: 0 ns, total: 6.07 ms
Wall time: 6.72 ms


MultinomialNB()

In [ ]:
# fazendo a predicao de x_test_dtm
y_pred_class = nb.predict(x_test_dtm)

In [ ]:
# calcular a acuracia da classe de predicao
from sklearn import metrics
metrics.accuracy_score(y_test,y_pred_class)

0.9856424982053122

In [ ]:
# print a matriz de confusao (contigencia)
metrics.confusion_matrix(y_test,y_pred_class)

array([[1205,    8],
       [  12,  168]])

<strong>1205</strong> - eram spam e foram classificados como spam<br />
<strong>8</strong>    - não eram spam, mas foram classicados como spam<br />
<strong>12</strong>   - eram spam, mas não foram classificados como spam<br />
<strong>168</strong>  - não eram spam, e não foram classificados como spam

In [ ]:
# imprima a mensagem como falso positivo 
# (o que significa que eles foram classificados incorretamente como spam)
x_test[y_test < y_pred_class]

325                      No calls..messages..missed calls
4598              Have you laid your airtel line to rest?
1289    Hey...Great deal...Farm tour 9am to 5pm $95/pa...
45                       No calls..messages..missed calls
573                                Waiting for your call.
3373                              Also andros ice etc etc
1081                    Can u get pic msgs to your phone?
494                      Are you free now?can i call now?
Name: message, dtype: object

In [ ]:
# imprima a mensagem como falso negativo 
# (o que significa que eles foram classificados incorretamente como nao spam)
x_test[y_test > y_pred_class]

4674    Hi babe its Chloe, how r u? I was smashed on s...
3528    Xmas & New Years Eve tickets are now on sale f...
1499    SMS. ac JSco: Energy is high, but u may not kn...
3417    LIFE has never been this much fun and great un...
2773    How come it takes so little time for a child w...
5       FreeMsg Hey there darling it's been 3 week's n...
1457    CLAIRE here am havin borin time & am now alone...
2429    Guess who am I?This is the first time I create...
4067    TBS/PERSOLVO. been chasing us since Sept forå£...
3358    Sorry I missed your call let's talk when you h...
2821    ROMCAPspam Everyone around should be respondin...
2247    Back 2 work 2morro half term over! Can U C me ...
Name: message, dtype: object

**Referência**:

Este notebook foi criado pelo prof. **Geanderson Esteves dos Santos** [UFMG].

